File: packages/benchmarks/src/metrics.ts
import type { RetrievalCase, RetrievalReport } from "./types.js";

export function normalizeAnswer(input: string): string {
  return input
    .toLowerCase()
    .replace(/\b(a|an|the)\b/g, " ")
    .replace(/[^\p{L}\p{N}]+/gu, " ")
    .trim()
    .replace(/\s+/g, " ");
}

export function tokenF1(predicted: string | undefined, expected: string | undefined): number {
  if (!predicted || !expected) return 0;
  const p = normalizeAnswer(predicted).split(" ").filter(Boolean);
  const e = normalizeAnswer(expected).split(" ").filter(Boolean);
  if (p.length === 0 || e.length === 0) return p.length === e.length ? 1 : 0;
  const counts = new Map<string, number>();
  for (const token of e) counts.set(token, (counts.get(token) ?? 0) + 1);
  let overlap = 0;
  for (const token of p) {
    const n = counts.get(token) ?? 0;
    if (n > 0) {
      overlap += 1;
      counts.set(token, n - 1);
    }
  }
  if (overlap === 0) return 0;
  const precision = overlap / p.length;
  const recall = overlap / e.length;
  return (2 * precision * recall) / (precision + recall);
}

function dcg(binaryRelevance: number[]): number {
  return binaryRelevance.reduce((sum, rel, i) => sum + rel / Math.log2(i + 2), 0);
}

function evaluateAtK(cases: RetrievalCase[], k: number) {
  const totals = { recall: 0, precision: 0, mrr: 0, ndcg: 0 };
  for (const c of cases) {
    const relevant = new Set(c.relevantIds);
    const top = c.retrievedIds.slice(0, k);
    const gains: number[] = [];
    const seenRelevant = new Set<string>();
    for (const id of top) {
      if (relevant.has(id) && !seenRelevant.has(id)) {
        seenRelevant.add(id);
        gains.push(1);
      } else {
        gains.push(0);
      }
    }
    const hits = seenRelevant.size;
    totals.recall += relevant.size === 0 ? 1 : hits / relevant.size;
    totals.precision += k === 0 ? 0 : hits / k;
    const firstHit = gains.findIndex((rel) => rel === 1);
    totals.mrr += firstHit === -1 ? 0 : 1 / (firstHit + 1);
    const idealHits = Math.min(relevant.size, k);
    const ideal = Array.from({ length: k }, (_, i) => (i < idealHits ? 1 : 0));
    const idealDcg = dcg(ideal);
    totals.ndcg += idealDcg === 0 ? 1 : dcg(gains) / idealDcg;
  }
  const n = Math.max(cases.length, 1);
  return {
    recall: totals.recall / n,
    precision: totals.precision / n,
    mrr: totals.mrr / n,
    ndcg: totals.ndcg / n,
  };
}

export function evaluateRetrieval(cases: RetrievalCase[], kValues: number[] = [1, 3, 5, 10]): RetrievalReport {
  const byK: RetrievalReport["byK"] = {};
  for (const k of kValues) byK[k] = evaluateAtK(cases, k);
  return { queryCount: cases.length, byK };
}

export function exactMatch(predicted: string | undefined, expected: string | undefined): number {
  if (!predicted || !expected) return 0;
  return normalizeAnswer(predicted) === normalizeAnswer(expected) ? 1 : 0;
}

File: go/internal/bench/metrics.go
package bench

// Scoring, transcribed from packages/benchmarks/src/metrics.ts. The
// arithmetic order is preserved because published numbers were produced
// by it — see docs/benchmarks/.

import (
	"math"
	"regexp"
	"sort"
	"strings"
)

var (
	articles    = regexp.MustCompile(`\b(a|an|the)\b`)
	nonAlnum    = regexp.MustCompile(`[^\p{L}\p{N}]+`)
	whitespaces = regexp.MustCompile(`\s+`)
)

// NormalizeAnswer lowercases, drops articles, strips punctuation, and
// collapses whitespace — SQuAD-style normalisation.
func NormalizeAnswer(input string) string {
	s := strings.ToLower(input)
	s = articles.ReplaceAllString(s, " ")
	s = nonAlnum.ReplaceAllString(s, " ")
	s = strings.TrimSpace(s)
	return whitespaces.ReplaceAllString(s, " ")
}

// ExactMatch is 1 when the normalised strings are identical.
func ExactMatch(predicted, expected string) float64 {
	if predicted == "" || expected == "" {
		return 0
	}
	if NormalizeAnswer(predicted) == NormalizeAnswer(expected) {
		return 1
	}
	return 0
}

// TokenF1 is the harmonic mean of token precision and recall, counting
// duplicate tokens only as often as they appear in the expectation.
func TokenF1(predicted, expected string) float64 {
	if predicted == "" || expected == "" {
		return 0
	}
	p := fields(NormalizeAnswer(predicted))
	e := fields(NormalizeAnswer(expected))
	if len(p) == 0 || len(e) == 0 {
		if len(p) == len(e) {
			return 1
		}
		return 0
	}
	counts := map[string]int{}
	for _, token := range e {
		counts[token]++
	}
	overlap := 0
	for _, token := range p {
		if counts[token] > 0 {
			overlap++
			counts[token]--
		}
	}
	if overlap == 0 {
		return 0
	}
	precision := float64(overlap) / float64(len(p))
	recall := float64(overlap) / float64(len(e))
	return (2 * precision * recall) / (precision + recall)
}

func fields(s string) []string {
	if s == "" {
		return nil
	}
	return strings.Split(s, " ")
}

func dcg(gains []float64) float64 {
	sum := 0.0
	for i, rel := range gains {
		sum += rel / math.Log2(float64(i)+2)
	}
	return sum
}

// RetrievalCase is the slice of a Case the retrieval metrics need.
type RetrievalCase struct {
	QueryID      string
	RelevantIDs  []string
	RetrievedIDs []string
}

func evaluateAtK(cases []RetrievalCase, k int) AtK {
	var totals AtK
	for _, c := range cases {
		relevant := map[string]bool{}
		for _, id := range c.RelevantIDs {
			relevant[id] = true
		}
		top := c.RetrievedIDs
		if len(top) > k {
			top = top[:k]
		}
		gains := make([]float64, 0, len(top))
		seen := map[string]bool{}
		for _, id := range top {
			if relevant[id] && !seen[id] {
				seen[id] = true
				gains = append(gains, 1)
			} else {
				gains = append(gains, 0)
			}
		}
		hits := len(seen)
		if len(relevant) == 0 {
			totals.Recall += 1
		} else {
			totals.Recall += float64(hits) / float64(len(relevant))
		}
		if k != 0 {
			totals.Precision += float64(hits) / float64(k)
		}
		firstHit := -1
		for i, rel := range gains {
			if rel == 1 {
				firstHit = i
				break
			}
		}
		if firstHit >= 0 {
			totals.MRR += 1 / float64(firstHit+1)
		}
		idealHits := len(relevant)
		if k < idealHits {
			idealHits = k
		}
		ideal := make([]float64, k)
		for i := range ideal {
			if i < idealHits {
				ideal[i] = 1
			}
		}
		idealDcg := dcg(ideal)
		if idealDcg == 0 {
			totals.NDCG += 1
		} else {
			totals.NDCG += dcg(gains) / idealDcg
		}
	}
	n := float64(len(cases))
	if n < 1 {
		n = 1
	}
	return AtK{
		Recall:    totals.Recall / n,
		Precision: totals.Precision / n,
		MRR:       totals.MRR / n,
		NDCG:      totals.NDCG / n,
	}
}

// EvaluateRetrieval scores every cutoff in kValues.
func EvaluateRetrieval(cases []RetrievalCase, kValues []int) RetrievalReport {
	byK := map[string]AtK{}
	for _, k := range kValues {
		byK[itoa(k)] = evaluateAtK(cases, k)
	}
	return RetrievalReport{QueryCount: len(cases), ByK: byK}
}

// percentile matches the TypeScript helper: sort ascending, then index
// ceil(p/100 * n) - 1, clamped to the last element.
func percentile(values []float64, p float64) float64 {
	if len(values) == 0 {
		return 0
	}
	sorted := append([]float64(nil), values...)
	sort.Float64s(sorted)
	idx := int(math.Ceil(p/100*float64(len(sorted)))) - 1
	if idx > len(sorted)-1 {
		idx = len(sorted) - 1
	}
	if idx < 0 {
		idx = 0
	}
	return sorted[idx]
}
